### ISSUE TITLE
Controlling stream representation

### ISSUE THREAD

**Issue by samhurst** [SpencerDawkins/sdp-roq#7]
This is related to my questions in #6, but this is something I've been experimenting on so wanted to surface it somewhere, and as this may result in some suggestions that go back to the RoQ draft itself I thought I won't detract from discussion there with this.

In my implementation, I've made it so the way that RTP frames are payloaded into QUIC streams can be controlled in the following ways:

- All RTP frames on a single stream (until the QUIC stream runs out of sending credit)
- All RTP frames for one or more GOPs on a single stream
- All RTP frames for one or more frames on a single stream

I control this with a "stream boundary" and "stream packing" parameter. The "stream boundary" parameter can be set to single, gop or frame, and then the "stream packing" parameter sets the number of GOPs or frames to put in each stream.

The reason you might want to do this is to be able to send `STOP_SENDING` or reset a stream you're no longer interested in, to get the sender to catch up to the latest frames.

The RoQ draft doesn't say anything on this subject, merely leaves it up to implementations to decide. I wonder whether this is something that we might want to specify more concretely, and where would be the best place to put it? Is there a use-case for a caller requesting a stream (i.e. `a=recvonly`) and then specifying how it wants that stream to be delivered? I've added `roq-stream-boundary` and `roq-stream-packing` attributes to my WIP implementation and have been experimenting with it a little bit with the intention of seeing whether it can positively affect latency on received streams.

**Comment by SpencerDawkins** [comment]
@samhurst - I THINK [Section 5.2.2 of the current RoQ specification](https://mengelbart.github.io/rtp-over-quic-draft/draft-ietf-avtcore-rtp-over-quic.html#name-media-frame-cancellation) addresses the way a receiver can get the sender to catch up to the current media frames. Does it look that way to you?

**Comment by samhurst** [comment]
Hi @SpencerDawkins ,

Section 5.2.2 certainly describes the mechanism for the receiver to stop the reception of frames on individual streams, and it's up to the sender to decide what that means, as the wording of 5.2.2 is not prescriptive over what the intended response by the sender should be in this instance. The only prescriptive part of it is that any media frame already sent on the QUIC stream that the receiver has stopped reading can't be sent again.

However, my question is more to do with controlling how the sender chooses to use streams to present RTP frames to the receiver, not in response to the receiver requesting that a given stream stops sending, although there is some overlap. Section 5.2 of the RoQ draft describes that RoQ senders may choose to do this to avoid head-of-line blocking. What I'm seeking to achieve here is how do we describe this behaviour in SDP?

**Comment by SpencerDawkins** [comment]
Hi, @samhurst, so, is the idea that the sender is telling the receiver what to expect, or that the receiver is telling the sender what to do? 

- If the sender is telling the receiver what to expect, that seems like a local decision at the sender (I know what is best for you). That might be more appropriate in RoQ, to make sure that receivers can handle reasonable sender behaviors. If that's what we're talking about, maybe we don't need to describe it in SDP. 
- if the receiver is telling the sender what to do, that absolutely needs to be in SDP, but I suspect that the behavior also needs to be in the RoQ draft. 

Am I following you closely enough? 

@mengelbart, do you have thoughts about this?

**Comment by samhurst** [comment]
In a SIP/SDP Offer/Answer world, the caller and callee don't necessarily map to sender and receiver of media. In the case where the media sender is also creating an SDP document (i.e. SIP caller or SIP callee creating a negotiated SDP document), I agree that it's probably not required to describe the media representation but I don't see the harm in including it. I can see a use case where the SDP will result in a new QUIC connection, you could use this information to help drive transport parameters. For example, for a sender sending on a single stream, it's peer could negotiate a very large `initial_max_stream_data_uni` transport parameter.

These are implementation decisions, so I don't think they should be in the RoQ draft which mostly deals with how RTP messages are carried on QUIC transport. RoQ doesn't have any real way of negotiating or signalling this kind of thing, and I would argue that it doesn't need it anyway, but I'm interested in what other people think.

**Comment by mengelbart** [comment]
I agree that RoQ doesn't need to signal this, but there may be cases where it could be useful for a receiver to know what to expect in order to provide sufficient flow control limits. `initial_max_stream_data_uni` is one example, `initial_max_streams_uni` is another example, which may be particularly important in large sessions with many streams where each frame opens a new stream.

I am not sure if this needs to be negotiated in SDP, though. I think in most cases, the sender will be in a better position to decide whether to open new streams or not and receivers should be prepared for that by providing enough flow control credits to allow the sender to do the right thing.

**Comment by SpencerDawkins** [comment]
I agree with @mengelbart. I'd add that signaling this means receivers need to manage flow control credits AND tell senders what to do, which seems unhelpful, if managing flow control credits alone is sufficient. I'll close this as wontfix.
